benchmarks/benchmark_concurrent_map.cpp

```diff
@@ -0,0 +1,192 @@
+static constexpr auto kThousand = 1'000;
+
+// Initially insert some numbers
+// Many threads check for contain
+//
+static auto ConcurrentManyContainsQueries(benchmark::State& state) -> void {
+  auto gen = std::mt19937{std::random_device{}()};
+  auto dis = std::uniform_int_distribution{-1 * kThousand, 1 * kThousand};
+
+  if (state.thread_index == 0) {
+    concurrent = std::make_unique<SL<int, int>>();
+    for (auto i = 0; i < kThousand * kThousand; ++i) {
+      auto num = dis(gen);
+      concurrent->Insert(num, num);
+    }
+  }
+
+  for (auto _ : state) {
+    auto num = dis(gen);
+    concurrent->Contains(num);
+  }
+
+  if (state.thread_index == 0) {
+    concurrent.reset();
+  }
+}
+
+// Initially insert some numbers
+// One thread inserts, others check for contain
+//
+static auto ConcurrentInitialOneInsertManyContainsQueries(
+    benchmark::State& state) -> void {
+  auto gen = std::mt19937{std::random_device{}()};
+  auto dis = std::uniform_int_distribution{-1 * kThousand, 1 * kThousand};
+
+  if (state.thread_index == 0) {
+    concurrent = std::make_unique<SL<int, int>>();
+    for (auto i = 0; i < kThousand * kThousand; ++i) {
+      auto num = dis(gen);
+      concurrent->Insert(num, num);
+    }
+  }
+
+  for (auto _ : state) {
+    auto num = dis(gen);
+    if (state.thread_index == 0) {
+      concurrent->Insert(num, num);
+    } else {
+      concurrent->Contains(num);
+    }
+  }
+
+  if (state.thread_index == 0) {
+    concurrent.reset();
+  }
+}
+
+// No initial setup
+// One thread inserts, others check for contain
+//
+static auto ConcurrentOneInsertManyContainsQueries(benchmark::State& state)
+    -> void {
+  auto gen = std::mt19937{std::random_device{}()};
+  auto dis = std::uniform_int_distribution{-1 * kThousand, 1 * kThousand};
+
+  if (state.thread_index == 0) {
+    concurrent = std::make_unique<SL<int, int>>();
+  }
+
+  for (auto _ : state) {
+    auto num = dis(gen);
+    if (state.thread_index == 0) {
+      concurrent->Insert(num, num);
+    } else {
+      concurrent->Contains(num);
+    }
+  }
+
+  if (state.thread_index == 0) {
+    concurrent.reset();
+  }
+}
+
+// Initially insert some numbers
+// One thread inserts, second (if exists) erases, others (if exist) check for
+// contain, every action in a loop
+//
+static auto ConcurrentOneInsertOneEraseManyContainsQueries(
+    benchmark::State& state) -> void {
+  auto gen = std::mt19937{std::random_device{}()};
+  auto dis = std::uniform_int_distribution{-1 * kThousand, 1 * kThousand};
+
+  if (state.thread_index == 0) {
+    concurrent = std::make_unique<SL<int, int>>();
+    for (auto i = 0; i < kThousand * kThousand; ++i) {
+      auto num = dis(gen);
+      concurrent->Insert(num, num);
+    }
+  }
+
+  for (auto _ : state) {
+    if (state.thread_index == 0) {
+      for (auto i = 0; i < kThousand; ++i) {
+        auto num = dis(gen);
+        concurrent->Insert(num, num);
+      }
+    } else if (state.threads > 1 && state.thread_index == 1) {
```

Comment:
`state.threads > 1` could be removed since `state.thread_index` will return `0` 
and the whole condition will be `false`.